politeiawww/user/mysqldb/mysqldb.go

```diff
@@ -0,0 +1,1018 @@
+		hex.EncodeToString(util.Digest([]byte(sid))))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SessionsDeleteByUserID deletes all sessions for the given user ID, except
+// the session IDs in exemptSessionIDs.
+//
+// SessionsDeleteByUserID satisfies the Database interface.
+func (m *mysqldb) SessionsDeleteByUserID(uid uuid.UUID, exemptSessionIDs []string) error {
+	log.Tracef("SessionsDeleteByUserID: %v %v", uid.String(), exemptSessionIDs)
+
+	ctx, cancel := ctxWithTimeout()
+	defer cancel()
+
+	// Session primary key is a SHA256 hash of the session ID.
+	exempt := make([]string, 0, len(exemptSessionIDs))
+	for _, v := range exemptSessionIDs {
+		exempt = append(exempt, hex.EncodeToString(util.Digest([]byte(v))))
+	}
+
+	// Using an empty NOT IN() set will result in no records being
+	// deleted.
+	if len(exempt) == 0 {
+		_, err := m.userDB.
+			ExecContext(ctx, "DELETE FROM sessions WHERE userID = ?", uid.String())
+		return err
+	}
+
+	_, err := m.userDB.
+		ExecContext(ctx, "DELETE FROM sessions WHERE usedID = ? AND key NOT IN (?)",
+			uid.String(), exempt)
+	return err
+}
+
+// RegisterPlugin registers a plugin.
+func (m *mysqldb) RegisterPlugin(p user.Plugin) error {
+	log.Tracef("RegisterPlugin: %v %v", p.ID, p.Version)
+
+	if m.isShutdown() {
+		return user.ErrShutdown
+	}
+
+	// Setup plugin tables
+	var err error
+	switch p.ID {
+	case user.CMSPluginID:
+		// XXX add the following:
+		// err = m.cmsPluginSetup()
+	default:
+		return user.ErrInvalidPlugin
+	}
+	if err != nil {
+		return err
+	}
+
+	// Save plugin settings.
+	m.Lock()
+	defer m.Unlock()
+
+	m.pluginSettings[p.ID] = p.Settings
+
+	return nil
+}
+
+// RotateKeys rotates the existing database encryption key with the given new
+// key.
+func (m *mysqldb) RotateKeys(newKeyPath string) error {
+	log.Tracef("RotateKeys: %v", newKeyPath)
+
+	if m.isShutdown() {
+		return user.ErrShutdown
+	}
+
+	// Load and validate new encryption key.
+	newKey, err := util.LoadEncryptionKey(log, newKeyPath)
+	if err != nil {
+		return fmt.Errorf("load encryption key '%v': %v",
+			newKeyPath, err)
+	}
+
+	if bytes.Equal(newKey[:], m.encryptionKey[:]) {
+		return fmt.Errorf("keys are the same")
+	}
+
+	log.Infof("Rotating encryption keys")
+
+	ctx, cancel := ctxWithTimeout()
+	defer cancel()
+
+	m.Lock()
+	defer m.Unlock()
+
+	// Rotate keys using a transaction.
+	opts := &sql.TxOptions{
+		Isolation: sql.LevelDefault,
+	}
+	tx, err := m.userDB.BeginTx(ctx, opts)
+	defer tx.Rollback()
+	if err != nil {
+		return err
+	}
+	err = rotateKeys(ctx, tx, m.encryptionKey, newKey)
+	if err != nil {
+		return err
+	}
+
+	// Commit transaction.
+	if err := tx.Commit(); err != nil {
+		return err
+	}
+
+	// Update context.
+	m.encryptionKey = newKey
+
+	return nil
+}
+
+// New connects to a mysql instance using the given connection params,
+// and returns pointer to the created mysql struct.
+func New(host, password, network, encryptionKey string) (*mysqldb, error) {
```

Comment:
This is a weird place for the `New` function.  Traditionally you put it either right under the struct that its initalizing or underneath all of the struct method definitions.  Putting it in the middle of the struct method defintions interupts the flow of the code.

Reply:
oh, good to know 👍🏽